### DISCUSSION TITLE
Recommended approach for populating Section component of a Form from an Action

### DISCUSSION THREAD

**Discussion by wickham-c** [filamentphp/filament#13670] [Help]
### Package

Form builder

### Package Version

v3.2.52

### How can we help you?

I've created a Form which comprises an "**Inputs**" Section component and an empty "**Outputs**" Section component.

The "**Inputs**" section contains a number of form fields for inputting data, plus a `Filament\Forms\Components\Actions\Action` component. In the closure for this action, I would like to populate the "**Outputs**" section with some HTML data (constructed using data from the aforementioned form fields) or better still, a Blade view.

I have no problem manipulating the form fields, but the part I am struggling with is how to push the HTML or Blade view into the empty "**Outputs**" Section component. 

So my question is:  _What would be the simplest or best way of achieving this?_

Any thoughts, hints. links or code snippets gratefully received.

**Comment by leandrocfe** [comment]
Share some code that you are using please

**Comment by wickham-c** [comment]
Certainly. Here's my **migration** (/database/migrations/2024_07_23_140903_create_examples_table.php)...
```
return new class extends \Illuminate\Database\Migrations\Migration
{
    public function up(): void
    {
        \Illuminate\Support\Facades\Schema::create('examples', function (\Illuminate\Database\Schema\Blueprint $table) {
            $table->id();
            $table->double('val1')->nullable();
            $table->double('val2')->nullable();
        });
    }
    ...
};
```
...and my **model** (/app/Models/Example.php)...
```
namespace App\Models;

class Example extends \Illuminate\Database\Eloquent\Model
{
    protected $fillable = [
        'val1',
        'val2',
    ];
}

```
...and my **resource** (/app/Filament/Resources/ExampleResource.php)...
```
namespace App\Filament\Resources;

use Filament\Forms;

class ExampleResource extends \Filament\Resources\Resource
{
    protected static ?string $model = \App\Models\Example::class;

    protected static ?string $navigationIcon = 'heroicon-o-rectangle-stack';

    public static function form(Forms\Form $form): Forms\Form
    {
        return $form
            ->schema([
                Forms\Components\Section::make('Inputs')
                    ->columns(2)
                    ->schema([
                        Forms\Components\TextInput::make('val1')
                            ->label('Operand 1')
                            ->numeric(),
                        Forms\Components\TextInput::make('val2')
                            ->label('Operand 2')
                            ->numeric(),
                        Forms\Components\Actions::make([
                            Forms\Components\Actions\Action::make('calculate')
                                 ->icon('fas-calculator')
                                 ->color('info')
                                 ->action(function (Forms\Set $set, $state) {
                                    $results = self::calculate_array_of_results($state['val1'], $state['val2']);
                                    $renderedView = \Illuminate\Support\Facades\Blade::render('results-output', ['results' => $results]);
                                    $set('Outputs', $renderedView);
                                 })
                             ])
                    ]),
                Forms\Components\Section::make('Outputs')
                    ->schema([
                        //
                    ])
            ]);
    }

    private static function calculate_array_of_results(float $val1, float $val2): array
    {
        return [
            'input1' => $val1,
            'input2' => $val2,
            'summation' => $val1 + $val2,
            'product' => $val1 * $val2,
            'average' => ($val1 + $val2) / 2,
            'range' => $val2 - $val1, 
        ];
    }
    ...
}
```
...and my **view** (/resources/views/results-output.blde.php)...
```
<table>
    <thead>
        <tr><th>Description</th><th>Result</th>
    </thead>
    <tbody>
        @foreach ($results as $key => $value)
            <tr><td>{{ $key }}</td><td>{{ $value }}</td></tr>
        @endforeach
    </tbody>
</table>
```

**Reply by leandrocfe** [reply]
I believe there are no issues if you are using a custom action to calculate custom attributes in your view.

**Comment by wickham-c** [comment]
Thanks @leandrocfe, but how do I actually render that view in the "**Outputs**" Section component?

**Reply by leandrocfe** [reply]
Did you try `->view('your-view)` ?

**Reply by wickham-c** [reply]
Thanks for the quick response @leandrocfe. I didn't know that method existed on a `Section` and, after a bit of digging, I can see that the `->view()` method is available on all components. 

Although, it would appear that it overwrites/replaces the view that draws the `Section` itself rather than pushing a view _inside_ the `Section` component.

**Reply by leandrocfe** [reply]
Use a view field or custom field: https://filamentphp.com/docs/3.x/forms/fields/custom#view-fields
You can set the `state` to your custom field `$set('your_custom_field', $results)` instead of using a custom view.
Inside your view field, you can get the value using `{{ $getState() }}`

**Reply by leandrocfe** [reply]
Then, you can use your custom field in your section

```php
Forms\Components\Section::make('Outputs')
->schema([
ViewField::make('your_custom_field')
    ->view('custom-view')
])
```
